File: app/src/main/java/com/waz/zclient/pages/main/MainPhoneFragment.java
/**
 * Wire
 * Copyright (C) 2016 Wire Swiss GmbH
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package com.waz.zclient.pages.main;

import android.content.Intent;
import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import com.waz.api.ErrorsList;
import com.waz.api.Message;
import com.waz.api.User;
import com.waz.model.MessageData;
import com.waz.zclient.BaseActivity;
import com.waz.zclient.OnBackPressedListener;
import com.waz.zclient.R;
import com.waz.zclient.controllers.accentcolor.AccentColorObserver;
import com.waz.zclient.controllers.collections.CollectionsObserver;
import com.waz.zclient.controllers.confirmation.ConfirmationObserver;
import com.waz.zclient.controllers.confirmation.ConfirmationRequest;
import com.waz.zclient.controllers.confirmation.IConfirmationController;
import com.waz.zclient.controllers.giphy.GiphyObserver;
import com.waz.zclient.controllers.navigation.Page;
import com.waz.zclient.controllers.singleimage.SingleImageObserver;
import com.waz.zclient.conversation.CollectionController;
import com.waz.zclient.conversation.CollectionFragment;
import com.waz.zclient.conversation.ShareToMultipleFragment;
import com.waz.zclient.core.stores.inappnotification.SyncErrorObserver;
import com.waz.zclient.fragments.ImageFragment;
import com.waz.zclient.pages.BaseFragment;
import com.waz.zclient.pages.main.conversation.SingleImageUserFragment;
import com.waz.zclient.pages.main.conversationlist.ConfirmationFragment;
import com.waz.zclient.pages.main.conversationpager.ConversationPagerFragment;
import com.waz.zclient.pages.main.giphy.GiphySharingPreviewFragment;
import com.waz.zclient.utils.SyncErrorUtils;
import com.waz.zclient.utils.ViewUtils;
import com.waz.zclient.views.menus.ConfirmationMenu;
import net.hockeyapp.android.ExceptionHandler;

public class MainPhoneFragment extends BaseFragment<MainPhoneFragment.Container> implements OnBackPressedListener,
                                                                                            ConversationPagerFragment.Container,
                                                                                            SingleImageObserver,
                                                                                            SingleImageUserFragment.Container,
                                                                                            GiphyObserver,
                                                                                            ConfirmationObserver,
                                                                                            AccentColorObserver,
                                                                                            CollectionsObserver,
                                                                                            ConfirmationFragment.Container,
    SyncErrorObserver {

    public static final String TAG = MainPhoneFragment.class.getName();
    private ConfirmationMenu confirmationMenu;

    ////////////////////////////////////////////////////////////////////////////////////////////////////////
    //
    //  LifeCycle
    //
    ////////////////////////////////////////////////////////////////////////////////////////////////////////

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_main, container, false);

        if (savedInstanceState == null) {
            getChildFragmentManager().beginTransaction().replace(R.id.fl_fragment_main_content,
                                                                 ConversationPagerFragment.newInstance(),
                                                                 ConversationPagerFragment.TAG).commit();
        }

        confirmationMenu = ViewUtils.getView(view, R.id.cm__confirm_action_light);
        confirmationMenu.setVisibility(View.GONE);

        return view;
    }

    @Override
    public void onStart() {
        super.onStart();
        getStoreFactory().inAppNotificationStore().addInAppNotificationObserver(this);
        getControllerFactory().getSingleImageController().addSingleImageObserver(this);
        getControllerFactory().getGiphyController().addObserver(this);
        getControllerFactory().getConfirmationController().addConfirmationObserver(this);
        getControllerFactory().getAccentColorController().addAccentColorObserver(this);
        getCollectionController().addObserver(this);
    }

    @Override
    public void onStop() {
        getControllerFactory().getGiphyController().removeObserver(this);
        getControllerFactory().getSingleImageController().removeSingleImageObserver(this);
        getControllerFactory().getConfirmationController().removeConfirmationObserver(this);
        getControllerFactory().getAccentColorController().removeAccentColorObserver(this);
        getStoreFactory().inAppNotificationStore().removeInAppNotificationObserver(this);
        getCollectionController().removeObserver(this);

        super.onStop();
    }

    @Override
    public void onDestroyView() {
        confirmationMenu = null;
        super.onDestroyView();
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        getChildFragmentManager().findFragmentById(R.id.fl_fragment_main_content).onActivityResult(requestCode,
                                                                                                   resultCode,
                                                                                                   data);
    }

    private CollectionController getCollectionController() {
        return ((BaseActivity) getActivity()).injectJava(CollectionController.class);
    }

    ////////////////////////////////////////////////////////////////////////////////////////////////////////
    //
    //  Notifications
    //
    ////////////////////////////////////////////////////////////////////////////////////////////////////////


    @Override
    public boolean onBackPressed() {
        if (confirmationMenu.getVisibility() == View.VISIBLE) {
            confirmationMenu.animateToShow(false);
            return true;
        }

        if (getChildFragmentManager().getBackStackEntryCount() > 0) {
            Fragment topFragment = getChildFragmentManager().findFragmentByTag(getChildFragmentManager().getBackStackEntryAt(
                getChildFragmentManager().getBackStackEntryCount() - 1).getName());
            if (topFragment instanceof SingleImageUserFragment) {
                return ((SingleImageUserFragment) topFragment).onBackPressed();
            } else if (topFragment instanceof GiphySharingPreviewFragment) {
                if (!((GiphySharingPreviewFragment) topFragment).onBackPressed()) {
                    getChildFragmentManager().popBackStackImmediate(GiphySharingPreviewFragment.TAG,
                                                                    FragmentManager.POP_BACK_STACK_INCLUSIVE);
                }
                return true;
            } else if (topFragment instanceof CollectionFragment) {
                return  ((CollectionFragment) topFragment).onBackPressed();
            } else if (topFragment instanceof ConfirmationFragment) {
                return ((ConfirmationFragment) topFragment).onBackPressed();
            } else if (topFragment instanceof ImageFragment) {
                if (!((ImageFragment) topFragment).onBackPressed()) {
                    getChildFragmentManager().popBackStackImmediate(ImageFragment.TAG(),
                        FragmentManager.POP_BACK_STACK_INCLUSIVE);
                }
                return true;
            }

        }

        // Back press is first delivered to the notification fragment, and if it's not consumed there,
        // it's then delivered to the main content.

        Fragment fragment = getChildFragmentManager().findFragmentById(R.id.fl_fragment_main_content);
        if (fragment instanceof OnBackPressedListener &&
            ((OnBackPressedListener) fragment).onBackPressed()) {
            return true;
        }

        fragment = getChildFragmentManager().findFragmentById(R.id.fl__overlay_container);
        if (fragment instanceof OnBackPressedListener &&
            ((OnBackPressedListener) fragment).onBackPressed()) {
            return true;
        }

        return getChildFragmentManager().popBackStackImmediate();
    }

    @Override
    public void onOpenUrl(String url) {
        getContainer().onOpenUrl(url);
    }

    @Override
    public void onShowSingleImage(Message message) {
        getChildFragmentManager().beginTransaction()
                                 .add(R.id.fl__overlay_container,
                                     ImageFragment.newInstance(message.getId()),
                                     ImageFragment.TAG())
                                 .addToBackStack(ImageFragment.TAG())
                                 .commit();
        getControllerFactory().getNavigationController().setRightPage(Page.SINGLE_MESSAGE, TAG);
    }

    @Override
    public void onShowUserImage(User user) {
        getChildFragmentManager().beginTransaction()
                                 .add(R.id.fl__overlay_container,
                                      SingleImageUserFragment.newInstance(user),
                                      SingleImageUserFragment.TAG)
                                 .addToBackStack(SingleImageUserFragment.TAG)
                                 .commit();
        getControllerFactory().getNavigationController().setRightPage(Page.SINGLE_MESSAGE, TAG);
    }

    @Override
    public void onHideSingleImage() {

    }

    @Override
    public void onSearch(String keyword) {
        getChildFragmentManager().beginTransaction()
                                 .add(R.id.fl__overlay_container,
                                      GiphySharingPreviewFragment.newInstance(keyword),
                                      GiphySharingPreviewFragment.TAG)
                                 .addToBackStack(GiphySharingPreviewFragment.TAG)
                                 .commit();
    }

    @Override
    public void onRandomSearch() {
        getChildFragmentManager().beginTransaction()
                                 .add(R.id.fl__overlay_container,
                                      GiphySharingPreviewFragment.newInstance(),
                                      GiphySharingPreviewFragment.TAG)
                                 .addToBackStack(GiphySharingPreviewFragment.TAG)
                                 .commit();
    }

    @Override
    public void onTrendingSearch() {
        getChildFragmentManager().beginTransaction()
                                 .add(R.id.fl__overlay_container,
                                      GiphySharingPreviewFragment.newInstance(),
                                      GiphySharingPreviewFragment.TAG)
                                 .addToBackStack(GiphySharingPreviewFragment.TAG)
                                 .commit();
    }

    @Override
    public void onCloseGiphy() {
        getChildFragmentManager().popBackStackImmediate(GiphySharingPreviewFragment.TAG,
                                                        FragmentManager.POP_BACK_STACK_INCLUSIVE);
    }

    @Override
    public void onCancelGiphy() {
        getChildFragmentManager().popBackStackImmediate(GiphySharingPreviewFragment.TAG,
                                                        FragmentManager.POP_BACK_STACK_INCLUSIVE);
    }

    //////////////////////////////////////////////////////////////////////////////////////////
    //
    //  CollectionsObserver
    //
    //////////////////////////////////////////////////////////////////////////////////////////

    @Override
    public void openCollection() {
    }

    @Override
    public void closeCollection() {
    }

    @Override
    public void shareCollectionItem(MessageData messageData) {
        getChildFragmentManager().beginTransaction()
                                .add(R.id.fl__overlay_container,
                                    ShareToMultipleFragment.newInstance(messageData.id()),
                                    ShareToMultipleFragment.TAG())
                                .addToBackStack(ShareToMultipleFragment.TAG())
                                .commit();
    }

    @Override
    public void closeCollectionShare() {
        getChildFragmentManager().popBackStackImmediate(ShareToMultipleFragment.TAG(),
            FragmentManager.POP_BACK_STACK_INCLUSIVE);
    }

    @Override
    public void nextItemRequested() {

    }

    @Override
    public void previousItemRequested() {

    }

    //////////////////////////////////////////////////////////////////////////////////////////
    //
    //  ConfirmationObserver
    //
    //////////////////////////////////////////////////////////////////////////////////////////

    @Override
    public void onRequestConfirmation(ConfirmationRequest confirmationRequest, @IConfirmationController.ConfirmationMenuRequester int requester) {
        confirmationMenu.onRequestConfirmation(confirmationRequest);
    }

    //////////////////////////////////////////////////////////////////////////////////////////
    //
    //  AccentColorObserver
    //
    //////////////////////////////////////////////////////////////////////////////////////////

    @Override
    public void onAccentColorHasChanged(Object sender, int color) {
        if (getView() == null) {
            return;
        }
        confirmationMenu.setButtonColor(color);
    }

    @Override
    public void onSyncError(ErrorsList.ErrorDescription error) {
        if (getActivity() == null) {
            return;
        }

        switch (error.getType()) {
            case CANNOT_ADD_UNCONNECTED_USER_TO_CONVERSATION:
            case CANNOT_ADD_USER_TO_FULL_CONVERSATION:
            case CANNOT_CREATE_GROUP_CONVERSATION_WITH_UNCONNECTED_USER:
                getChildFragmentManager().beginTransaction()
                                         .setCustomAnimations(R.anim.fade_in, R.anim.fade_out, R.anim.fade_in, R.anim.fade_out)
                                         .replace(R.id.fl_dialog_container,
                                                  ConfirmationFragment.newMessageOnlyInstance(getResources().getString(R.string.in_app_notification__sync_error__create_group_convo__title),
                                                                                              SyncErrorUtils.getGroupErrorMessage(getContext(), error),
                                                                                              getResources().getString(R.string.in_app_notification__sync_error__create_convo__button),
                                                                                              error.getId()),
                                                  ConfirmationFragment.TAG
                                                 )
                                         .addToBackStack(ConfirmationFragment.TAG)
                                         .commit();
                break;
            case CANNOT_ADD_USER_TO_FULL_CALL:
            case CANNOT_CALL_CONVERSATION_WITH_TOO_MANY_MEMBERS:
            case CANNOT_SEND_VIDEO:
            case PLAYBACK_FAILURE:
                ExceptionHandler.saveException(new RuntimeException("Unhandled error " + error.getType()), null);
                break;
            case CANNOT_SEND_MESSAGE_TO_UNVERIFIED_CONVERSATION:
            case RECORDING_FAILURE:
            case CANNOT_SEND_ASSET_FILE_NOT_FOUND:
            case CANNOT_SEND_ASSET_TOO_LARGE:
                // Handled in ConversationFragment
                break;
            default:
                ExceptionHandler.saveException(new RuntimeException("Unexpected error " + error.getType()), null);

        }
    }

    @Override
    public void onDialogConfirm(String dialogId) {
        getChildFragmentManager().popBackStackImmediate(ConfirmationFragment.TAG,
                                                        FragmentManager.POP_BACK_STACK_INCLUSIVE);
        dismissError(dialogId);
    }

    @Override
    public void onDialogCancel(String dialogId) {
        getChildFragmentManager().popBackStackImmediate(ConfirmationFragment.TAG,
                                                        FragmentManager.POP_BACK_STACK_INCLUSIVE);
        dismissError(dialogId);
    }

    private void dismissError(String errorId) {
        if (getActivity() == null ||
            getStoreFactory().isTornDown()) {
            return;
        }
        getStoreFactory().inAppNotificationStore().dismissError(errorId);
    }

    public interface Container {
        void onOpenUrl(String url);
    }
}

File: app/src/main/scala/com/waz/zclient/pages/main/MainPhoneFragment.scala
/**
 * Wire
 * Copyright (C) 2018 Wire Swiss GmbH
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package com.waz.zclient.pages.main

import android.content.Intent
import android.os.Bundle
import android.support.v4.app.FragmentManager
import android.view.{LayoutInflater, View, ViewGroup}
import com.waz.model.ErrorData
import com.waz.service.ZMessaging
import com.waz.threading.Threading
import com.waz.utils.events.Signal
import com.waz.utils.returning
import com.waz.zclient.collection.controllers.CollectionController
import com.waz.zclient.collection.fragments.CollectionFragment
import com.waz.zclient.common.controllers.global.AccentColorController
import com.waz.zclient.controllers.collections.CollectionsObserver
import com.waz.zclient.controllers.confirmation.{ConfirmationObserver, ConfirmationRequest}
import com.waz.zclient.controllers.giphy.GiphyObserver
import com.waz.zclient.controllers.navigation.Page
import com.waz.zclient.controllers.singleimage.SingleImageObserver
import com.waz.zclient.conversation.{ConversationController, ImageFragment}
import com.waz.zclient.messages.UsersController
import com.waz.zclient.pages.BaseFragment
import com.waz.zclient.pages.main.conversationlist.ConfirmationFragment
import com.waz.zclient.pages.main.conversationpager.ConversationPagerFragment
import com.waz.zclient.pages.main.giphy.GiphySharingPreviewFragment
import com.waz.zclient.views.menus.ConfirmationMenu
import com.waz.zclient.{FragmentHelper, OnBackPressedListener, R}
import net.hockeyapp.android.ExceptionHandler

import scala.concurrent.Future

/**
  * Wire
  * Copyright (C) 2018 Wire Swiss GmbH
  *
  * This program is free software: you can redistribute it and/or modify
  * it under the terms of the GNU General Public License as published by
  * the Free Software Foundation, either version 3 of the License, or
  * (at your option) any later version.
  *
  * This program is distributed in the hope that it will be useful,
  * but WITHOUT ANY WARRANTY; without even the implied warranty of
  * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  * GNU General Public License for more details.
  *
  * You should have received a copy of the GNU General Public License
  * along with this program.  If not, see <http://www.gnu.org/licenses/>.
  */

class MainPhoneFragment extends BaseFragment[MainPhoneFragment.Container]
  with FragmentHelper
  with OnBackPressedListener
  with ConversationPagerFragment.Container
  with SingleImageObserver
  with GiphyObserver
  with ConfirmationObserver
  with CollectionsObserver
  with ConfirmationFragment.Container {

  import MainPhoneFragment._
  import Threading.Implicits.Ui

  private lazy val zms = inject[Signal[ZMessaging]]

  private lazy val usersController = inject[UsersController]
  private lazy val conversationController = inject[ConversationController]
  private lazy val accentColorController = inject[AccentColorController]
  private lazy val collectionController = inject[CollectionController]

  private lazy val confirmationMenu = returning(view[ConfirmationMenu](R.id.cm__confirm_action_light)) { vh =>
    accentColorController.accentColor.onUi(color => vh.foreach(_.setButtonColor(color.getColor)))
  }

  override def onCreateView(inflater: LayoutInflater, container: ViewGroup, savedInstanceState: Bundle): View = {
    if (savedInstanceState == null)
      getChildFragmentManager
        .beginTransaction
        .replace(R.id.fl_fragment_main_content, ConversationPagerFragment.newInstance, ConversationPagerFragment.TAG)
        .commit

    inflater.inflate(R.layout.fragment_main, container, false)
  }

  override def onViewCreated(view: View, savedInstanceState: Bundle): Unit = {
    confirmationMenu.foreach(_.setVisibility(View.GONE))
    zms.flatMap(_.errors.getErrors).onUi { _.foreach(handleSyncError) }
  }

  override def onStart(): Unit = {
    super.onStart()
    getControllerFactory.getSingleImageController.addSingleImageObserver(this)
    getControllerFactory.getGiphyController.addObserver(this)
    getControllerFactory.getConfirmationController.addConfirmationObserver(this)
    collectionController.addObserver(this)
  }

  override def onStop(): Unit = {
    getControllerFactory.getGiphyController.removeObserver(this)
    getControllerFactory.getSingleImageController.removeSingleImageObserver(this)
    getControllerFactory.getConfirmationController.removeConfirmationObserver(this)
    collectionController.removeObserver(this)
    super.onStop()
  }

  override def onActivityResult(requestCode: Int, resultCode: Int, data: Intent): Unit = {
    super.onActivityResult(requestCode, resultCode, data)
    withFragment(R.id.fl_fragment_main_content)(_.onActivityResult(requestCode, resultCode, data))
  }

  override def onBackPressed(): Boolean = confirmationMenu flatMap { confirmationMenu =>
    if (confirmationMenu.getVisibility == View.VISIBLE) {
      confirmationMenu.animateToShow(false)
      return true
    }

    val backStackSize = getChildFragmentManager.getBackStackEntryCount
    lazy val topFragment = getChildFragmentManager.findFragmentByTag(getChildFragmentManager.getBackStackEntryAt(backStackSize - 1).getName)
    val mainContentFragment = getChildFragmentManager.findFragmentById(R.id.fl_fragment_main_content)
    val overlayContentFragment = getChildFragmentManager.findFragmentById(R.id.fl__overlay_container)

    if (backStackSize > 0) {
      Option(topFragment) collect {
        case f : GiphySharingPreviewFragment =>
          f.onBackPressed() || getChildFragmentManager.popBackStackImmediate(GiphySharingPreviewFragment.TAG, FragmentManager.POP_BACK_STACK_INCLUSIVE)
        case f : ImageFragment =>
          f.onBackPressed() || getChildFragmentManager.popBackStackImmediate(ImageFragment.Tag, FragmentManager.POP_BACK_STACK_INCLUSIVE)
        case f : CollectionFragment => f.onBackPressed()
        case f : ConfirmationFragment => f.onBackPressed()
      }
    } else {
      // Back press is first delivered to the notification fragment, and if it's not consumed there,
      // it's then delivered to the main content.
      Option(mainContentFragment) collect {
        case f : OnBackPressedListener if f.onBackPressed() => true
      } orElse (Option(overlayContentFragment) collect {
        case f : OnBackPressedListener if f.onBackPressed() => true
      })
    }

  } getOrElse getChildFragmentManager.popBackStackImmediate

  override def onOpenUrl(url: String): Unit = getContainer.onOpenUrl(url)

  override def onShowSingleImage(messageId: String): Unit = {
    getChildFragmentManager
      .beginTransaction
      .add(R.id.fl__overlay_container, ImageFragment.newInstance(messageId), ImageFragment.Tag)
      .addToBackStack(ImageFragment.Tag)
      .commit
    getControllerFactory.getNavigationController.setRightPage(Page.SINGLE_MESSAGE, Tag)
  }

  override def onHideSingleImage(): Unit = ()

  override def onSearch(keyword: String): Unit = openGiphyPreviewFragment()

  override def onRandomSearch(): Unit = openGiphyPreviewFragment()

  override def onTrendingSearch(): Unit = openGiphyPreviewFragment()

  override def onCloseGiphy(): Unit = closeGiphyPreviewFragment()

  override def onCancelGiphy(): Unit = closeGiphyPreviewFragment()

  private def openGiphyPreviewFragment(): Unit = {
    import GiphySharingPreviewFragment._
    getChildFragmentManager
      .beginTransaction
      .add(R.id.fl__overlay_container, newInstance, TAG)
      .addToBackStack(TAG)
      .commit
  }

  private def closeGiphyPreviewFragment(): Unit = {
    import GiphySharingPreviewFragment._
    getChildFragmentManager.popBackStackImmediate(TAG, FragmentManager.POP_BACK_STACK_INCLUSIVE)
  }

  override def openCollection(): Unit = ()

  override def closeCollection(): Unit = ()

  override def onRequestConfirmation(confirmationRequest: ConfirmationRequest, requester: Int): Unit = {
    confirmationMenu.onRequestConfirmation(confirmationRequest)
  }

  private def handleSyncError(error: ErrorData): Unit = {
    import ConfirmationFragment._
    import com.waz.ZLog.ImplicitTag._
    import com.waz.api.ErrorType._

    def getGroupErrorMessage: Future[String] = {
      error.errType match {
        case CANNOT_ADD_UNCONNECTED_USER_TO_CONVERSATION =>
          if (error.users.size == 1)
            usersController.user(error.users.head).head
              .map(getString(R.string.in_app_notification__sync_error__add_user__body, _))
          else
            Future.successful(getString(R.string.in_app_notification__sync_error__add_multiple_user__body))
        case CANNOT_CREATE_GROUP_CONVERSATION_WITH_UNCONNECTED_USER =>
          conversationController.conversationData(error.convId.get).head
            .map(data => getString(R.string.in_app_notification__sync_error__create_group_convo__body, data.get.displayName))
        case _ =>
          Future.successful(getString(R.string.in_app_notification__sync_error__unknown__body))
      }
    }

    error.errType match {
      case CANNOT_ADD_UNCONNECTED_USER_TO_CONVERSATION |
           CANNOT_ADD_USER_TO_FULL_CONVERSATION |
           CANNOT_CREATE_GROUP_CONVERSATION_WITH_UNCONNECTED_USER =>
        getGroupErrorMessage foreach { errorMsg =>
          getChildFragmentManager
            .beginTransaction
            .setCustomAnimations(R.anim.fade_in, R.anim.fade_out, R.anim.fade_in, R.anim.fade_out)
            .replace(
              R.id.fl_dialog_container,
              newMessageOnlyInstance(
                getResources.getString(R.string.in_app_notification__sync_error__create_group_convo__title),
                errorMsg,
                getResources.getString(R.string.in_app_notification__sync_error__create_convo__button),
                error.id.str
              ),
              TAG
            )
            .addToBackStack(TAG)
            .commit
        }
      case CANNOT_ADD_USER_TO_FULL_CALL |
           CANNOT_CALL_CONVERSATION_WITH_TOO_MANY_MEMBERS |
           CANNOT_SEND_VIDEO |
           PLAYBACK_FAILURE =>
        ExceptionHandler.saveException(new RuntimeException("Unhandled error " + error.errType), null, null)
      case CANNOT_SEND_MESSAGE_TO_UNVERIFIED_CONVERSATION |
           RECORDING_FAILURE |
           CANNOT_SEND_ASSET_FILE_NOT_FOUND |
           CANNOT_SEND_ASSET_TOO_LARGE => // Handled in ConversationFragment
      case _ =>
        ExceptionHandler.saveException(new RuntimeException("Unexpected error " + error.errType), null, null)
    }
  }

  override def onDialogConfirm(dialogId: String): Unit = closeConfirmationDialog(dialogId)

  override def onDialogCancel(dialogId: String): Unit = closeConfirmationDialog(dialogId)

  private def closeConfirmationDialog(dialogId: String): Unit = {
    getChildFragmentManager.popBackStackImmediate(ConfirmationFragment.TAG, FragmentManager.POP_BACK_STACK_INCLUSIVE)
    dismissError(dialogId)
  }

  private def dismissError(errorId: String) = {
    if (getActivity != null && !getStoreFactory.isTornDown)
      getStoreFactory.inAppNotificationStore.dismissError(errorId)
  }
}

object MainPhoneFragment {
  val Tag: String = classOf[MainPhoneFragment].getName

  trait Container {
    def onOpenUrl(url: String): Unit
  }

}

